JAVA-3168 Copy node info for contact points on initial node refresh only from first match by endpoint

jahstreet opened the pull request:
https://datastax-oss.atlassian.net/browse/JAVA-3168

DataStax Java Cassandra driver supports[ FixedHostNameAddressTranslator](https://github.com/apache/cassandra-java-driver/blob/4.17.0/core/src/main/java/com/datastax/oss/driver/internal/core/addresstranslation/FixedHostNameAddressTranslator.java) since[ v4.15.0](https://docs.datastax.com/en/developer/java-driver/4.17/changelog/index.html#4-15-0) (Sep 19, 2022). This [address translator](https://docs.datastax.com/en/developer/java-driver/4.17/manual/core/address_resolution/index.html#driver-side-address-translation) plugin allows Cassandra clients to connect to a Cassandra database running in a different (private) network, eg. Kubernetes, via load balancer.

On the[ initial (control) connection](https://docs.datastax.com/en/developer/java-driver/4.17/manual/core/control_connection/index.html) to Cassandra, the driver queries the DB cluster for its topology and fetches its nodes IDs together with their IP addresses (Pod IPs in a private K8s network). These addresses are "translated" by the address translator to the configured URL pointing to the load balancer. So the client "thinks" it has connected to all Cassandra nodes, however the node addresses are the same.

Connecting to Cassandra behind load balancer with FixedHostNameAddressTranslatorAfter, the driver opens a connection pool to every discovered Cassandra node. And given all the node addresses now point to the load balancer, all the connections are getting opened to it. We can calculate the total number of connections as:

`client_app_count * cassandra_node_count * connections_per_node_count`

By default, the driver sets `connections_per_node_count` to 1 ([advanced.connection.pool](https://docs.datastax.com/en/developer/java-driver/4.17/manual/core/configuration/reference/index.html)), however for data intensive applications it is configured to a higher value. For example, [Apache Spark Cassandra Connector](https://github.com/datastax/spark-cassandra-connector) overrides it to the number of available JVM CPUs ([spark.cassandra.connection.localConnectionsPerExecutor](https://github.com/datastax/spark-cassandra-connector/blob/master/doc/reference.md)). So it can be a usual scenario to have a Spark job with 64 executors (each running a separate instance of a Cassandra driver) with 16 CPU cores connecting to a 100 node Cassandra cluster, which generates `64 * 100 * 16 = 102'400` connections. And that is a single application. So we must be careful when configuring the load balancer to not exceed limits on number of connections, number of open files, memory, etc.

Driver, when fetched the cluster topology and translated node addresses, does "optimization": it creates an instance of Node class for every contact point address used on initial (control) connection and[ reuses them](https://github.com/apache/cassandra-java-driver/blob/4.17.0/core/src/main/java/com/datastax/oss/driver/internal/core/metadata/InitialNodeListRefresh.java#L75) for all the translated node addresses if they match any of the contact points. Then, in the scenario when contact point is the load balancer address and address translator also translates node addresses to the load balancer address, the[ driver metadata](https://github.com/apache/cassandra-java-driver/blob/4.17.0/core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultMetadata.java#L69C26-L69C34) will contain node IDs mapping to the same Node instance, eg.: `node_id_1 -> LB_Node, node_id_2 -> LB_Node, node_id_3 -> LB_Node, …`

Then the [connection pool is initialized for each node](https://github.com/apache/cassandra-java-driver/blob/4.17.0/core/src/main/java/com/datastax/oss/driver/internal/core/session/PoolManager.java#L231) (or better say multiple times for the same Node), but with a "bug": after a pool is created, it is[ put to a map with Node as a key](https://github.com/apache/cassandra-java-driver/blob/4.17.0/core/src/main/java/com/datastax/oss/driver/internal/core/session/PoolManager.java#L250), and when Node is the same instance for all pools (LB_Node), we get a map with a single pool in it (each time we put a new value for the same key, the value is being overwritten). So, in result we have created pools for every discovered Cassandra node, but all except one of them are leaked and will be running in a background keeping the connections alive, but staying unused.

With that change, the driver reuses contact point Nodes and fills them with the missing node info on initial node refresh only from first match with the node info by endpoint. That ensures we always create separate Node instances for Nodes having the same EndPoints, which down the line protects us from leaking the connection (channel) pools in PoolManager.

```
patch by Alex Sasnouskikh; reviewed by Andy Tolbert and Alexandre Dura for JAVA-3168
```

tolbertam commented on core/src/main/java/com/datastax/oss/driver/internal/core/session/PoolManager.java:
```diff
@@ -221,18 +222,23 @@ private void init(CqlIdentifier keyspace) {
 
       Collection<Node> nodes = context.getMetadataManager().getMetadata().getNodes().values();
-      List<CompletionStage<ChannelPool>> poolStages = new ArrayList<>(nodes.size());
+      Map<Node, CompletionStage<ChannelPool>> poolStageByNode = new HashMap<>(nodes.size());
```
One doubt I have about this is that it's dependent on the `Node` implementation.  It looks like `DefaultNode` doesn't implement either `equals` or `hashCode` so I suspect it will treat every node as different, even if it has the same underlying `EndPoint`.  Instead we may need to key this by `EndPoint`.

In the case of `should_not_connect_to_same_nodes`, it seems to work because the same reference is used.

jahstreet replied:
Indeed, it deduplicates `DefaultNode`s by reference, which is also the case down the line in https://github.com/apache/cassandra-java-driver/pull/2007/files#diff-79645c264b633a5c25cadd4cfd581e8ae06fb41a1adc5f51087ccc46ce0d7452R258.
Then we look up pools by node instance as well on requests in `DefaultSession#getChannel`.
No matter the node implementation, we need to make sure we do not `leak` pools by losing them on `pools#put` with the same key.
On the other side, if we key by `EndPoint`, then we also need to get by `EndPoint`, cause otherwise on requests we risk to not find the desired pool for the filtered out node instance.
WDYT?

tolbertam replied:
Wrote a [test](https://gist.github.com/tolbertam/fe877d03623a6f0392153a5493c1bfb2) to demonstrate reproducing this, unfortunately I didn't get a chance to validate the pool behavior:

![image](https://github.com/user-attachments/assets/95113342-9894-4ee3-ba9e-d0019fc59360)

> On the other side, if we key by EndPoint, then we also need to get by EndPoint, cause otherwise on requests we risk to not find the desired pool for the filtered out node instance.

Yeah I think that is inherently the problem.  It's not clear to me why currently it's not actively creating connections for each `Node`.  There's something else I need to understand there so I'm still debugging to better understand.

tolbertam replied:
I see, looks like there is a problem with my test where it throws an UnknownHostException because i'm configuring this wrong. Fixing this up quick:

```
10:05:31.474 [s0-admin-1] WARN  c.d.o.d.i.core.pool.ChannelPool - [s0|/127.0.0.1:9043]  Error while opening new channel
java.net.UnknownHostException: /127.0.0.1
	at java.base/java.net.InetAddress$CachedAddresses.get(InetAddress.java:797)
	at java.base/java.net.InetAddress.getAllByName0(InetAddress.java:1533)
	at java.base/java.net.InetAddress.getAllByName(InetAddress.java:1386)
	at java.base/java.net.InetAddress.getAllByName(InetAddress.java:1307)
	at java.base/java.net.InetAddress.getByName(InetAddress.java:1257)
	at io.netty.util.internal.SocketUtils$8.run(SocketUtils.java:156)
	at io.netty.util.internal.SocketUtils$8.run(SocketUtils.java:153)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at io.netty.util.internal.SocketUtils.addressByName(SocketUtils.java:153)
	at io.netty.resolver.DefaultNameResolver.doResolve(DefaultNameResolver.java:41)
	at io.netty.resolver.SimpleNameResolver.resolve(SimpleNameResolver.java:61)
	at io.netty.resolver.SimpleNameResolver.resolve(SimpleNameResolver.java:53)
	at io.netty.resolver.InetSocketAddressResolver.doResolve(InetSocketAddressResolver.java:55)
	at io.netty.resolver.InetSocketAddressResolver.doResolve(InetSocketAddressResolver.java:31)
	at io.netty.resolver.AbstractAddressResolver.resolve(AbstractAddressResolver.java:106)
	at io.netty.bootstrap.Bootstrap.doResolveAndConnect0(Bootstrap.java:206)
	at io.netty.bootstrap.Bootstrap.access$000(Bootstrap.java:46)
	at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:180)
	at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:166)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:590)
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:557)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:492)
	at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:636)
	at io.netty.util.concurrent.DefaultPromise.setSuccess0(DefaultPromise.java:625)
	at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:105)
	at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:84)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.safeSetSuccess(AbstractChannel.java:990)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:516)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:429)
	at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:486)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute$$$capture(AbstractEventExecutor.java:167)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

tolbertam replied:
I updated the test to properly provide `advanced.address-translator.advertised-hostname` and encountered a new issue while building the token map:

```
10:08:12.788 [s0-admin-0] WARN  c.d.o.d.i.c.metadata.DefaultMetadata - [s0] Unexpected error while refreshing token map, keeping previous version
java.lang.IllegalArgumentException: Multiple entries with same key: Murmur3Token(3074457345618258602)=Node(endPoint=/127.0.0.1:9043, hostId=36c0b4f5-08dc-4872-8dbc-85242f72568c, hashCode=4194c04f) and Murmur3Token(3074457345618258602)=Node(endPoint=/127.0.0.1:9043, hostId=36c0b4f5-08dc-4872-8dbc-85242f72568c, hashCode=4194c04f)
	at com.datastax.oss.driver.shaded.guava.common.collect.ImmutableMap.conflictException(ImmutableMap.java:382)
	at com.datastax.oss.driver.shaded.guava.common.collect.ImmutableMap.checkNoConflict(ImmutableMap.java:376)
	at com.datastax.oss.driver.shaded.guava.common.collect.RegularImmutableMap.checkNoConflictInKeyBucket(RegularImmutableMap.java:249)
	at com.datastax.oss.driver.shaded.guava.common.collect.RegularImmutableMap.fromEntryArrayCheckingBucketOverflow(RegularImmutableMap.java:136)
	at com.datastax.oss.driver.shaded.guava.common.collect.RegularImmutableMap.fromEntryArray(RegularImmutableMap.java:98)
	at com.datastax.oss.driver.shaded.guava.common.collect.ImmutableMap$Builder.build(ImmutableMap.java:579)
	at com.datastax.oss.driver.shaded.guava.common.collect.ImmutableMap$Builder.buildOrThrow(ImmutableMap.java:607)
	at com.datastax.oss.driver.shaded.guava.common.collect.ImmutableMap$Builder.build(ImmutableMap.java:594)
	at com.datastax.oss.driver.internal.core.metadata.token.DefaultTokenMap.buildTokenToPrimaryAndRing(DefaultTokenMap.java:263)
	at com.datastax.oss.driver.internal.core.metadata.token.DefaultTokenMap.build(DefaultTokenMap.java:59)
	at com.datastax.oss.driver.internal.core.metadata.DefaultMetadata.rebuildTokenMap(DefaultMetadata.java:161)
	at com.datastax.oss.driver.internal.core.metadata.DefaultMetadata.withNodes(DefaultMetadata.java:117)
	at com.datastax.oss.driver.internal.core.metadata.InitialNodeListRefresh.compute(InitialNodeListRefresh.java:106)
	at com.datastax.oss.driver.internal.core.metadata.MetadataManager.apply(MetadataManager.java:519)
	at com.datastax.oss.driver.internal.core.metadata.MetadataManager$SingleThreaded.refreshNodes(MetadataManager.java:335)
	at com.datastax.oss.driver.internal.core.metadata.MetadataManager$SingleThreaded.access$1700(MetadataManager.java:299)
	at com.datastax.oss.driver.internal.core.metadata.MetadataManager.lambda$refreshNodes$1(MetadataManager.java:172)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
	at io.netty.channel.DefaultEventLoop.run(DefaultEventLoop.java:54)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

looking through the `InitialNodeListRefresh` code to see if there is other issue that is reliant on the endpoint instead of host id.

tolbertam replied:
Ah, the issue is here in `InitialNodeListRefresh.compute()`

```
        EndPoint endPoint = nodeInfo.getEndPoint();
        DefaultNode node = findIn(contactPoints, endPoint);
        if (node == null) {
          node = new DefaultNode(endPoint, context);
          LOG.debug("[{}] Adding new node {}", logPrefix, node);
        } else {
          LOG.debug("[{}] Copying contact point {}", logPrefix, node);
        }
```

jahstreet replied:
You might also wanna check https://github.com/apache/cassandra-java-driver/blob/4.17.0/core/src/main/java/com/datastax/oss/driver/internal/core/metadata/InitialNodeListRefresh.java#L75-L85, specifically `#copyInfo` which messes up the Node properties.
On the token info, there is no reason to keep it enabled when all nodes point to the same address (which is the case for the FixedHostNameAddressTranslator), which is most probably a load balancer routing the requests on its own logic. We should somehow hint users on it though, agree, not nice to just fail. Would you recommend to handle it in code or document somehow?
So, try to disable token info, it should move you further.

tolbertam replied:
Only reference I could find on this is something I stumbled into in 2019: https://lists.apache.org/thread/91cvsrg33tgsqh9jhqckxkhsn3z6zy7j

This code also seems problematic and I think we should probably do away with it in favor of creating a new `Node` every time instead of copying it from the contact point.

tolbertam replied:
Making this change:

```diff
--- a/core/src/main/java/com/datastax/oss/driver/internal/core/metadata/InitialNodeListRefresh.java
+++ b/core/src/main/java/com/datastax/oss/driver/internal/core/metadata/InitialNodeListRefresh.java
@@ -74,13 +74,8 @@ class InitialNodeListRefresh extends NodesRefresh {
             hostId);
       } else {
         EndPoint endPoint = nodeInfo.getEndPoint();
-        DefaultNode node = findIn(contactPoints, endPoint);
-        if (node == null) {
-          node = new DefaultNode(endPoint, context);
-          LOG.debug("[{}] Adding new node {}", logPrefix, node);
-        } else {
-          LOG.debug("[{}] Copying contact point {}", logPrefix, node);
-        }
+        DefaultNode node = new DefaultNode(endPoint, context);
+        LOG.debug("[{}] Adding new node {}", logPrefix, node);
         if (tokenMapEnabled && tokenFactory == null && nodeInfo.getPartitioner() != null) {
           tokenFactory = tokenFactoryRegistry.tokenFactoryFor(nodeInfo.getPartitioner());
         }
```

I'd probably change this code further to reutilize the node if the host id and token were the same.

I see that connections are established for each node, and my test now fails (which is what I expected).

I think if we were to change the logic such that we key pools by `EndPoint`, my test would now pass.

Are you in ASF slack?  There's a [thread](https://the-asf.slack.com/archives/C05LPRVNZV1/p1738160135983149) for this issue here that Santosh Pingale started that maybe we can chat more interactively in there

jahstreet replied:
I'm not, but he promised to invite me ;)
Or can I ask you for it please ? My Slack is on jahstreetlove@gmail.com.

jahstreet replied:
Adding a node each time is also a good option 👍 . Lets chat to see what works best.

tolbertam replied:
sounds good, sent you an invite and looks like it was already approved :+1:

tolbertam commented on core/src/main/java/com/datastax/oss/driver/internal/core/session/PoolManager.java:
```diff
@@ -221,18 +222,23 @@ private void init(CqlIdentifier keyspace) {
         } else if (node.getState() == NodeState.FORCED_DOWN) {
           LOG.debug("[{}] Skipping {} because it is FORCED_DOWN", logPrefix, node);
+        } else if (poolStageByNode.containsKey(node)) {
+          LOG.warn("[{}] Skipping {} because pool for it is already created", logPrefix, node);
```
Probably sufficient to use debug here :+1:
```suggestion
          LOG.debug("[{}] Skipping {} because pool for it is already created", logPrefix, node);
```

jahstreet replied:
Ofc, agree. Changing.

jahstreet commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/InitialNodeListRefresh.java:
```diff
@@ -90,30 +83,16 @@ public Result compute(
-      if (findIn(newNodes.values(), contactPoint.getEndPoint()) == null) {
-        eventsBuilder.add(NodeStateEvent.removed(contactPoint));
-      }
+      eventsBuilder.add(NodeStateEvent.removed(contactPoint));
```
Will this change force control connection to always reconnect on initial node refresh?

jahstreet replied:
Yes, it happens in com.datastax.oss.driver.internal.core.control.ControlConnection.SingleThreaded#onStateEvent .

jahstreet replied:
These docs are useful to grasp it high level https://docs.datastax.com/en/developer/java-driver/4.15/manual/developer/admin/index.html

jahstreet replied:
What if we check whether to reuse the contact point node by host ID? Then we can avoid that.

jahstreet replied:
Nope, contact point nodes come without host ID, they are built in `com.datastax.oss.driver.internal.core.session.DefaultSession.SingleThreaded#init` -> `com.datastax.oss.driver.internal.core.metadata.MetadataManager#addContactPoints`. We can try to do first match by `EndPoint` though.

tolbertam reviewed (approved):
Looks great!  Excellent work!

adutra commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/InitialNodeListRefresh.java:
```diff
@@ -63,22 +63,29 @@ public Result compute(
     Map<UUID, DefaultNode> newNodes = new HashMap<>();
+    // Contact point nodes don't have host ID as well as other info yet, so we fill them with node
+    // info found on first match by endpoint
+    Set<DefaultNode> matchedContactPoints = new HashSet<>();
```
I'm a bit surprised: `DefaultNode` doesn't implement `hashCode()` and `equals()`. I wonder if methods like `contains()` and `add()` (used below) won't misbehave? Wouldn't be better to use `Set<EndPoint>` here? It seems doable.

tolbertam replied:
ah right interesting, didn't occur to me at the time, but made a similar observation on the [initial commit](https://github.com/apache/cassandra-java-driver/pull/2007#discussion_r1934067152).  I suspect that for some reason this still works as I still see `Adding new node {}` in my test, double checking this though.

tolbertam replied:
yeah, it does seem to work out, debugging through why though:

```
12:03:36.493 [s0-admin-0] DEBUG c.d.o.d.i.c.m.InitialNodeListRefresh - [s0] Copying contact point Node(endPoint=/127.0.0.1:9043, hostId=null, hashCode=78f4e2c6)
12:03:36.493 [s0-admin-0] DEBUG c.d.o.d.i.c.m.t.DefaultTokenFactoryRegistry - [s0] Detected Murmur3 partitioner (org.apache.cassandra.dht.Murmur3Partitioner)
12:03:36.493 [s0-admin-0] DEBUG c.d.o.d.i.c.m.InitialNodeListRefresh - [s0] Adding new node Node(endPoint=/127.0.0.1:9043, hostId=null, hashCode=5b31d39e)
12:03:36.493 [s0-admin-0] DEBUG c.d.o.d.i.c.m.InitialNodeListRefresh - [s0] Adding new node Node(endPoint=/127.0.0.1:9043, hostId=null, hashCode=15439812)
```

tolbertam replied:
Oh right, because `findContactPointNode` grabs the `Node` from `contactPoints` and adds it to `matchedContactPoints` .  On each iteration `findContactPointNode` grabs the exact object, so `matchedContactPoints.contains(contactPointNode)` evaluates to true after the first encountered one.

adutra replied:
Yeah that's what I suspected. It only works because the objects are the same. But still, that's a bit fragile. I think using `Set<EndPoint>` would also work, and would be more future-proof.

tolbertam replied:
agreed :+1:

jahstreet replied:
Hi @adutra 👋 .
Thanks for the review.
I actually relied on equality of references here, so that was intentional. I can look into refactoring to use EndPoint, from the first PoV it should work same. If we assume that contact points are always linked to the different EndPoints all will be smooth.

jahstreet replied:
I see we store contact point EndPoints in Set when creating Nodes for them in `com.datastax.oss.driver.internal.core.metadata.MetadataManager#addContactPoints`, so we should be good to assume their uniqueness.

jahstreet replied:
Pushed a commit to address the proposal, @adutra please take a look 🙏 .
I like it less, it requires adding `addedNodes` collection and down the line we still rely on DefaultNode equality checks by reference, eg. in PoolManager. But I'm also good to go as-is.
If you approve, I will squash the commits and add the required commit description about the author and reviewers (linked it in the footer of the PR description).

adutra replied:
Looking awesome @jahstreet thanks for your contribution!

tolbertam reviewed (approved):
:+1:

jahstreet committed:
JAVA-3168 Copy node info for contact points on initial node refresh only from first match by endpoint

patch by Alex Sasnouskikh; reviewed by Andy Tolbert and Alexandre Dura for JAVA-3168

jahstreet commented:
Squashed commits, PTAL.
Ready to go 🚀

tolbertam commented:
This is great, thank you @jahstreet !
